File: internal/index/video_scene/video_scene.go
package videoscene

import (
	"strings"
	"sync"

	"github.com/syndtr/goleveldb/leveldb"
)

type VideoSceneRecord struct {
	VideoID  string
	SceneIDs []string
}

type RecordVisitor interface {
	Visit(VideoSceneRecord) error
}

type VideoSceneIndex interface {
	InsertVideoScenes(videoID string, sceneIds []string)
	ForEach(RecordVisitor) error
}

type LevelVideoSceneIndex struct {
	db *leveldb.DB
}

func (i *LevelVideoSceneIndex) InsertVideoScenes(videoID string, sceneIds []string) {
	i.db.Put([]byte(videoID), []byte(strings.Join(sceneIds, ",")), nil)
}

func (i *LevelVideoSceneIndex) ForEach(vis RecordVisitor) error {
	iter := i.db.NewIterator(nil, nil)
	wg := &sync.WaitGroup{}

	for iter.Next() {
		key := iter.Key()
		value := iter.Value()

		r := VideoSceneRecord{
			VideoID:  string(key),
			SceneIDs: strings.Split(string(value), ","),
		}

		wg.Add(1)
		go func(g *sync.WaitGroup) {
			vis.Visit(r)
			wg.Done()
		}(wg)
		// err := vis.Visit(r)
		// if err != nil {
		// 	return err
		// }
	}
	wg.Wait()
	return nil
}

func New(path string) (VideoSceneIndex, error) {
	db, err := leveldb.OpenFile(path, nil)
	if err != nil {
		return nil, err
	}

	return &LevelVideoSceneIndex{
		db: db,
	}, nil
}

File: indexes/video_scene.py
import typing
import leveldb

class VideoSceneRecord():
    def __init__(self, videoID: str, sceneIDs: typing.List[str]):
        self.videoID = videoID
        self.sceneIDs = sceneIDs


class VideoSceneIndex():
    def __init__(self, path) -> None:
        self.level_instance = leveldb.LevelDB(path)

    def collect_records(self) -> typing.List[VideoSceneRecord]:
        r = []
        for k, v in self.level_instance.RangeIter():
            record = VideoSceneRecord(k.decode('utf-8'), str(v.decode('utf-8')).split(','))
            r.append(record)

        return r

    def insert(self, videoID, sceneIds: typing.List[str]):
        b = ",".join(sceneIds)
        self.level_instance.Put(videoID.encode('utf-8'), b.encode('utf-8'))